cocos/animation/marionette/animation-graph-animation-clip-binding.ts

```diff
@@ -0,0 +1,349 @@
+                    return segments.join('/');
+                } else {
+                    return undefined;
+                }
+            })();
+            if (typeof transformPath === 'string') {
+                const transformHandle = bindContext.bindTransform(transformPath);
+                if (!transformHandle) {
+                    return undefined;
+                }
+                return bindPoseTransform(transformHandle, lastPropertyKey);
+            }
+        }
+    }
+
+    // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    // TODO: here should be resolved before this can be landed.
+    error(`Animation graph currently only supports (bone) transform animations.`);
+    return undefined;
+}
+
+/**
+ * Describes the evaluation of a animation clip in sense of animation graph.
+ */
+export class AnimationClipAGEvaluation {
+    constructor (
+        clip: AnimationClip,
+        context: AnimationClipGraphBindingContext,
+    ) {
+        clip.__trySyncLegacyData();
+
+        const trackEvaluations: AGTrackEvaluation<unknown>[] = [];
+        let exoticAnimationEvaluation: ExoticTrsAGEvaluation | undefined;
+
+        const {
+            tracks,
+            [exoticAnimationTag]: exoticAnimation,
+        } = clip;
+
+        for (const track of tracks) {
+            if (track instanceof UntypedTrack) {
+            // Untyped track is not supported in AG.
+                continue;
+            }
+            if (Array.from(track.channels()).every(({ curve }) => curve.keyFramesCount === 0)) {
+                continue;
+            }
+            const trackRuntimeBinding = bindTrackAG(clip, track, context);
+            if (!trackRuntimeBinding) {
+                continue;
+            }
+            const trackSampler = track[createEvalSymbol]();
+            const trackEvaluation = new AGTrackEvaluation(trackRuntimeBinding, trackSampler);
+            trackEvaluations.push(trackEvaluation);
+        }
+
+        if (exoticAnimation) {
+            exoticAnimationEvaluation = exoticAnimation.createEvaluatorForAnimationGraph(context);
+        }
+
+        this._trackEvaluations = trackEvaluations;
+        this._exoticAnimationEvaluation = exoticAnimationEvaluation;
+    }
+
+    /**
+     * Destroys all the track evaluations and exotic animation evaluation.
+     */
+    public destroy () {
+        this._exoticAnimationEvaluation?.destroy();
+
+        this._trackEvaluations.forEach((trackEvaluation) => {
+            trackEvaluation.destroy();
+        });
+    }
+
+    /**
+     * Evaluates.
+     * @param time The time.
+     * @param context The evaluation context.
+     */
+    public evaluate (time: number, context: AnimationClipGraphEvaluationContext) {
+        const {
+            _trackEvaluations: trackEvaluations,
+            _exoticAnimationEvaluation: exoticAnimationEvaluation,
+        } = this;
+
+        const pose = context;
+
+        for (const trackEvaluation of trackEvaluations) {
```

Comment:
Avoid using `for of` in perframe operetion.  The compiled code has additional overhead.
![image](https://user-images.githubusercontent.com/15109777/209267634-4db27d04-11d3-4bab-bb9f-37739e76c25b.png)

Reply:
https://github.com/cocos/cocos-engine/pull/13503/commits/1780961fe0e8bd27cf0a5c145aecde1ef759ce97